### ISSUE TITLE
More flexible aggregations of effects? Allow for ranges? For simultaneous restrictions on groups and times?

### ISSUE THREAD

**Issue by jtorcasso** [bcallaway11/did#97]
There's already a lot of flexibility in the package for aggregating the individual 2x2 ATT(g,t)s. But I know I'd find it useful to give the user a little bit more control over the overall average. 

Ranges: Is there a way to specify a range over which you'd like the overall average? For instance, in event time, if you wanted to get the average impact for only event times 6 through 12. You may want to do this if you expect there to be a ramp up period where the treatment effect increases until stabilizing at event time 6. While you may want to stop at 12 to avoid changing the composition of your cohorts too much in event time.

Simultaneous restrictions: Suppose I want an overall average for the earlier treated groups, say g < 10, AND I want to restrict to a particular event time range, or calendar time range. 

How to do this: perhaps the easiest way is to have a new function that allows overall aggregations, where the user can supply up to two vectors, choosing among groups, calendar times, or event times.

**Comment by bcallaway11** [comment]
Yes, I am really interested in these sorts of issues --- i.e., making code generic enough that users can (fairly easily) come up with their own application-specific solutions.  

This is not exactly what you are asking for, but we have made a lot of progress on writing generic code:

- At the moment, the `att_gt` function can accommodate custom estimation methods through its `est_method` argument (this is documented [here](https://bcallaway11.github.io/did/reference/att_gt.html))
- I have another package, [pte](https://github.com/bcallaway11/pte), that aims for letting users implement alternative identification strategies (besides DID) when there is treatment effect heterogeneity and variation in treatment timing.

With regard to your specific comment, while it would not be super user-friendly at the moment, it should be possible to implement quite general custom aggregation schemes.  In particular, `att_gt` returns the influence function for the estimated ATT(g,t)'s, which means that you can often pretty easily come up with a custom aggregation.  One example is #58.

I'm going to leave this open and tag this as a possible enhancement going forward.

**Comment by jtorcasso** [comment]
Thanks for the pointers! For a workaround, I ended up modifying the compute.aggte function. For the dynamic effects, I changed the line `epos <- eseq >= 0` to `epos <- eseq >= overall_min_e & eseq <= overall_max_e`, after defining new parameter options `overall_min_e` and `overall_max_e`.

**Comment by ocolluphid** [comment]
@bcallaway11 if one wants to define a custom aggregation (e.g. comparing dynamic treatment effects of cohorts before 2020 vs. dynamic TEs of treated cohorts after 2020) would you recommend using the regular bootstrap or the influence functions for valid SEs? I imagine a naive bootstrap might seem straightforward but very time consuming. 

see a similar discussion I started in STATA forum here, which wasn't entirely resolved: https://www.statalist.org/forums/forum/general-stata-discussion/general/1782212-custom-aggregation-and-linear-contrasts-with-csdid

thanks in advance!

**Comment by pedrohcgs** [comment]
I would get the IFs or do a weighted(Bayesian) bootstrap. The reason is
that I am a bit concerned that some nonparametric bootstrap draws do not
have obs from a cohort. If that is not a concern, all these would be ok.

Since the package already export the IFs of the ATT(g,t)’s, maybe the extra
work is little



----------------------------------------------------
*Pedro H. C. Sant'Anna*
*https://psantanna.com <https://psantanna.com>*

----------------------------------------------------
Warning: This email may contain confidential or privileged information
intended only for the use of the individual or entity to whom it is
addressed. If you are not the intended recipient, please understand
that any disclosure, copying, distribution, or use of the contents
of this email is strictly prohibited.
----------------------------------------------------

**Comment by pedrohcgs** [comment]
I dont have anything off-the-shelve…
Thanks

----------------------------------------------------
*Pedro H. C. Sant'Anna*
*https://psantanna.com <https://psantanna.com>*

----------------------------------------------------
Warning: This email may contain confidential or privileged information
intended only for the use of the individual or entity to whom it is
addressed. If you are not the intended recipient, please understand
that any disclosure, copying, distribution, or use of the contents
of this email is strictly prohibited.
----------------------------------------------------

**Comment by pedrohcgs** [comment]
Linking this to #254, which tracks the implementation of `coef()` and `vcov()` S3 methods for `MP` and `AGGTEobj` objects.

Once those are in place, custom linear restrictions on ATT(g,t) estimates — including ranges of event times, specific cohort subsets, or joint restrictions across groups and times — become straightforward via `car::linearHypothesis()`:

```r
# e.g. test whether average effect for cohorts treated before 2020
# equals average effect for cohorts treated after 2020
car::linearHypothesis(mp, hypothesis.matrix = H)
```

The `min_e` / `max_e` arguments on `aggte()` already partially address the "ranges" part of this request for the dynamic aggregation case.

We will keep this open until #254 is resolved, at which point most of the use cases described here should be covered.
